arches/app/datatypes/url.py

```diff
@@ -82,6 +83,17 @@ def validate(self, value, row_number=None, source=None, node=None, nodeid=None,
                 title = _("Invalid HTTP/HTTPS URL")
                 error_message = self.create_error_message(value, source, row_number, message, title)
                 errors.append(error_message)
+            
+            # raise error if label added without URL (#10592)
+            if value.get("url_label") and not value.get("url"):
+                message = _("URL label cannot be saved without a URL")
+                title = _("No URL added")
+                error_message = self.create_error_message(value, source, row_number, message, title)
+                errors.append(error_message)
+
+            if "url_label" not in value:
```

Comment:
The validate method shouldn't modify data. It should only check if a value is valid. The pre_tile_save datatype method is intended for this kind of change.

Reply:
I noticed that the urldatatype.py didn't have a `pre_tile_save` function, so I have added one and moved the fix from validate.  Will reflect the changes in #10643

One thing I have noticed is that this fix helps to validate all future urldatatype additions, however, all of the url datatype tiles saved prior to this fix will still lack the `"url_label": ""` from their data, and thus the UI will not allow them to add a label unless the tile is deleted and data re-populated.

Is there some way we can check and correct all previously saved data that doesn't contain "url_label" and insert the key with an empty string value (`"url_label": ""`) like we are doing here?

Reply:
> and thus the UI will not allow them to add a label unless the tile is deleted and data re-populated.

Indeed, when I try to do this, I see two things:
1. No dirty state detected
2. TypeError:
```js
urldatatype.js:17 Uncaught TypeError: this.url_label is not a function
    at viewModel.eval (urldatatype.js:17:36)
```

I think we need to find a way to guard this on the frontend.

> Is there some way we can check and correct all previously saved data that doesn't contain "url_label" and insert the key with an empty string value ("url_label": "") like we are doing here?

We could add a data migration, but then we'd want to target 7.6. We should probably just be more permissive in the UI.

Reply:
Hey @SDScandrettKint I just wanted to make sure this comment wasn't lost. Do you have a view about how to proceed here? Should we guard for this in the UI and just create the missing key on demand?

Reply:
Thanks for bringing this to my attention, you were right it was lost - I somehow didn't see this comment.
I also see the lack of dirty state and an error in the browser console.

I've been looking at this for a few days, trying to get a dirty state in the UI, but this approach failed as it doesn't seem to be possible to turn this.tile.dirty(true).  Similarly, modifying this.url_label also didn't affect the dirty tile state.  After speaking with @chiatt I've changed the approach and added the `pre_structure_tile_data` function which uses the code from validate to add the "url_label" key and allow editing of the label.  Then upon save the key will be saved to the JSON.  Though this approach doesn't prompt a dirty state, it does avoid the need for a migration.

This change is in the latest commit [60c1af4](https://github.com/archesproject/arches/pull/10642/commits/60c1af4d61441bf1517777dc0ba9999a1260ec52)

Reply:
Great. I retested and dirty state works fine now, too.